File: java/150/EvaluateReversePolishNotation.java
class Solution {

    /** 使用集合来判断字符串是否为数字，代码可读性更好 */
    private static final Set<String> set = new HashSet<>(Arrays.asList("+", "-", "*", "/"));

    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            if (!set.contains(token)) {
                stack.push(Integer.parseInt(token));
            } else {
                int a = stack.pop();
                int b = stack.pop();
                if ("+".equals(token)) {
                    stack.push(b + a);
                } else if ("-".equals(token)) {
                    stack.push(b - a);
                } else if ("*".equals(token)) {
                    stack.push(b * a);
                } else {
                    stack.push(b / a);
                }
            }
        }
        return stack.pop();
    }
}

File: python/150/Evaluate_Reverse_Polish_Notation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        for x in tokens:
            if not x in "+-*/":
                stack.append(int(x))
            else:
                first = stack.pop()
                second = stack.pop()
                res = int(eval(str(second * 1.0) + x + str(first)))
                stack.append(res)
        return stack[-1]

# 6 / -132 = -1
# int(6.0 / -132) = 0
